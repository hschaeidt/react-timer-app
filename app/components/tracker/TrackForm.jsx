import React, { Component, PropTypes } from 'react';
import moment from 'moment';

export default class TrackForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      day: moment().format('YYYY-MM-DD'),
    };

    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleDayChange = this.handleDayChange.bind(this);
  }

  handleSubmit(e) {
    e.preventDefault();

    if (this.state.day.match(/^[0-9]{4}-[0-9]{2}-[0-9]{2}$/) && moment(this.state.day).isValid()) {
      this.props.onAddTrack(this.state.day);
    }
  }

  handleDayChange(e) {
    this.setState({
      day: e.target.value,
    });
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <input
          type="date"
          value={this.state.day}
          onChange={this.handleDayChange}
        />
        <button className="button expanded">Check-in</button>
      </form>
    );
  }
}

TrackForm.propTypes = {
  onAddTrack: PropTypes.func.isRequired,
};
