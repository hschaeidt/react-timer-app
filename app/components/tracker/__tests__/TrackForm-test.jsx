import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import TrackForm from '../TrackForm';

Date.now = jest.fn(() => 1480896000000);

function setup() {
  const props = {
    onAddTrack: jest.fn(),
  };

  const wrapper = mount(<TrackForm {...props} />);

  return {
    wrapper,
    props,
  };
}

describe('<TrackForm />', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<TrackForm onAddTrack={jest.fn()} />);
    expect(tree).toMatchSnapshot();
  });

  it('should call onAddTrack prop with valid data', () => {
    const { wrapper, props } = setup();

    wrapper.find('input').first().simulate('change', { target: { value: '2013-12-25' } });
    wrapper.simulate('submit');
    expect(props.onAddTrack).toHaveBeenCalledWith('2013-12-25');
  });

  it('should not call onAddTrack with data not matching yyyy-MM-dd', () => {
    const { wrapper, props } = setup();

    wrapper.find('input').first().simulate('change', { target: { value: 'abcdefg' } });
    wrapper.simulate('submit');
    expect(props.onAddTrack).not.toHaveBeenCalled();
  });
});
