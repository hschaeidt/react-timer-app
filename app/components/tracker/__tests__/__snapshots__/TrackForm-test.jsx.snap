exports[`<TrackForm /> renders correctly 1`] = `
<form
  onSubmit={[Function]}>
  <input
    onChange={[Function]}
    type="date"
    value="2016-12-05" />
  <button
    className="button expanded">
    Check-in
  </button>
</form>
`;
